uspace/drv/bus/usb/ehci/ehci_batch.c

```diff
@@ -348,38 +322,38 @@ static void batch_control(ehci_transfer_batch_t *ehci_batch, usb_direction_t dir
  * Direction is supplied by the associated ep and toggle is maintained by the
  * EHCI hw in ED.
  */
-static void batch_data(ehci_transfer_batch_t *ehci_batch, usb_direction_t dir)
+static void batch_data(ehci_transfer_batch_t *ehci_batch)
 {
 	assert(ehci_batch);
-	assert(ehci_batch->usb_batch);
-	assert(dir == USB_DIRECTION_IN || dir == USB_DIRECTION_OUT);
 
-	usb_log_debug2("Batch %p: Data QH(%"PRIxn"): "
-	    "%08x:%08x:%08x:%08x:%08x:%08x", ehci_batch->usb_batch,
-	    addr_to_phys(ehci_batch->qh),
+	usb_log_debug2("Batch %p: Data QH(%p): "
+	    "%08x:%08x:%08x:%08x:%08x:%08x", ehci_batch,
+	    ehci_batch->qh,
 	    ehci_batch->qh->ep_char, ehci_batch->qh->ep_cap,
 	    ehci_batch->qh->status, ehci_batch->qh->current,
 	    ehci_batch->qh->next, ehci_batch->qh->alternate);
 
 	size_t td_current = 0;
-	size_t remain_size = ehci_batch->usb_batch->buffer_size;
-	char *buffer = ehci_batch->device_buffer;
+	size_t remain_size = ehci_batch->base.size;
+	uintptr_t buffer = dma_buffer_phys(&ehci_batch->base.dma_buffer,
+	    ehci_batch->data_buffer);
 	while (remain_size > 0) {
 		const size_t transfer_size = remain_size > EHCI_TD_MAX_TRANSFER
 		    ? EHCI_TD_MAX_TRANSFER : remain_size;
 
 		const bool last = (remain_size == transfer_size);
-		td_init(
-		    ehci_batch->tds[td_current],
-		    last ? NULL : ehci_batch->tds[td_current + 1],
-		    dir, buffer, transfer_size, -1, last);
+		td_init(&ehci_batch->tds[td_current],
+		    last ? 0 : dma_buffer_phys(&ehci_batch->ehci_dma_buffer,
+			    &ehci_batch->tds[td_current + 1]),
```

Comment:
Nit: The last tab should be four spaces instead.

Reply:
Well, it's a continuation of a line continuation (more refactoring would certainly help!), so IMO it shall be 8 spaces.

Reply:
A continuation line does not have a continuation line of its own with a different indent. One big line can have multiple continuation lines that are each indented the same - by four spaces.

Reply:
Ok, my comment is not entirely clear if you don't have visible whitespace or a custom tab width. The line has three tab indents and four spaces indentation. It should be two tabs and eight spaces. IMO the extra level of continuation is correct, because the nested continuation continues a nested syntactic construct. Flat four space continution would harm readability. Or, you could use intermediate variables. That would probably be best.

Reply:
This not however how it is supposed to be used.